src/supernova/test.rs

```diff
@@ -466,55 +403,37 @@ where
 
   let mut recursive_snark_option: Option<RecursiveSNARK<G1, G2>> = None;
 
-  for _ in 0..num_steps {
-    let program_counter = recursive_snark_option.as_ref().map_or_else(
-      || initial_program_counter,
-      |recursive_snark| recursive_snark.program_counter,
-    );
-
-    // The program counter directly specifies the next circuit index.
-    let augmented_circuit_index = u32::from_le_bytes(
-      // convert program counter from field to usize (only took le 4 bytes)
-      program_counter.to_repr().as_ref()[0..4].try_into().unwrap(),
-    ) as usize;
+  for &op_code in test_rom.rom.iter() {
+    let mut recursive_snark = recursive_snark_option.unwrap_or_else(|| {
+      RecursiveSNARK::new(
+        &pp,
+        &test_rom,
+        &test_rom.primary_circuit(op_code),
+        &test_rom.secondary_circuit(),
+        &z0_primary,
+        &z0_secondary,
+      )
+      .unwrap()
+    });
 
-    let mut recursive_snark =
-      recursive_snark_option.unwrap_or_else(|| match augmented_circuit_index {
-        OPCODE_0 | OPCODE_1 => RecursiveSNARK::new(
+    let circuit_primary = test_rom.primary_circuit(op_code);
+    let circuit_secondary = test_rom.secondary_circuit();
```

Comment:
Maybe move these bindings to the top of the loop and use them in the binding of `recursive_snark` also.

Reply:
Good catch!